File: frontend/src/interfaces/room.ts
import type { FormValues } from "../lib/form";
import type { BuildingData } from "./building";
import type { RoomStatus } from "./enums";
import type { RoomTypeData } from "./roomType";

export interface RoomData {
	id: number;
	status: RoomStatus;
	description?: string;
	roomNumber: string;
	floorNumber: number;
	createdAt: Date;
	updatedAt?: Date;
	roomType: RoomTypeData;
	building: BuildingData;
}

export interface RoomFormData extends FormValues {
	status: RoomStatus;
	description: string;
	roomNumber: string;
	floorNumber: number;
	roomTypeId: number;
	buildingId: number;
}

export interface CreateRoomPayload {
	status: RoomStatus;
	description?: string;
	roomNumber: string;
	floorNumber: number;
	roomTypeId: number;
	buildingId: number;
}

export interface UpdateRoomPayload {
	status: RoomStatus;
	description?: string;
	roomNumber: string;
	floorNumber: number;
	roomTypeId?: number;
	buildingId?: number;
}

File: backend/src/main/java/com/hpms/backend/model/Room.java
package com.hpms.backend.model;

import com.hpms.backend.enumCollection.RoomStatusEnum;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.Collection;
import java.util.HashSet;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "rooms")
public class Room {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @Enumerated(EnumType.STRING)
    private RoomStatusEnum status;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private String roomNumber;

    private int floorNumber;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @ManyToOne
    @JoinColumn(name = "room_type_id", nullable = false)
    private RoomType roomType;

    @ManyToOne
    @JoinColumn(name = "building_id", nullable = false)
    private Building building;
    @OneToMany(mappedBy = "room", cascade = CascadeType.ALL, orphanRemoval = true)
    private Collection<Housekeeping> housekeepingTasks = new HashSet<>();
    @ManyToMany(mappedBy = "rooms")
    private Collection<Booking> bookings = new HashSet<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

}